google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -55,6 +55,9 @@ auto MakeLinks(Spans::const_iterator begin, Spans::const_iterator end) {
 
 auto MakeParent(Links const& links, Spans const& message_spans) {
   namespace sc = ::opentelemetry::trace::SemanticConventions;
+  // // Detach current otel context before creating the span.
+  internal::DetachOTelContext(
+      opentelemetry::context::RuntimeContext::GetCurrent());
```

Comment:
> Detaching here looks strange to me. It is either clever or a hack.

On second look, this is not needed. I removed it and it generated the same [spans](https://screenshot.googleplex.com/4sfkiCbW9Uiqd6L).
 
> Can you say which context (i.e. which span name) we are detaching in the comment? I do not remember which one is supposed to be current.

It's detaching the "x publish" span created in the PublisherTracingConnection.

> So we have N messages, and on the Nth one the batch is full so we flush it. We hit this code with the "publisher send" span active. Is that right?

Yes

> And the logic is that even though the code goes from message N to flushing the batch, we just don't think a parent-child relationship makes sense. It was all of the messages from 1-N that led to the batch flushing.

Yes correct, since we are using links to associate them

> Alternatively, we might want to detach all of these "publisher send" spans earlier if no work is ever done on their behalf. (If work is done on behalf of "publisher send" spans after the flush.... then we probably can't detach the "publisher send" context here...).

Reply:
Now, with the hack part removed, does this change sound good?

The alternative I think is possible based on what you described.

Add a SaveMessage API to Batch Sink, Move TracingMB to TracingBatchSink, and wrap it, but I think it will be 2-3 weeks of work for the same effect.

The only thing I'm not sure about is the BatchSink holds the CQ so we try to just make one. So I think it might be a little ugly but we want to do something like

```
    if (!key.empty()) {
          // Only wrap the sink if there is an ordering key.
    used_sink = pubsub_internal::TracingBatchSink::Create(std::move(pubsub_internal::SequentialBatchSink::Create(
              std::move(used_sink)))
        } else {
          used_sink = pubsub_internal::TracingBatchSink::Create(
              std::move(used_sink));
}
```
in publisher connection, since we only want add the seq batch sink.

Reply:
Also this might be a change we want to make after there exists an end to end integration test that verifies the spans https://github.com/googleapis/google-cloud-cpp/issues/13034

Reply:
> I think it will be 2-3 weeks of work for the same effect.

1. I am glad you have learned the ways of conservative estimates. I don't think it would be too bad, though.
2. Sure the effect would be the same, but the code to achieve that effect would be less complicated, and thus easier to maintain.

> I think it might be a little ugly

It can be less ugly.

```
```

Reply:
> > I think it will be 2-3 weeks of work for the same effect.
> 
> 1. I am glad you have learned the ways of conservative estimates. I don't think it would be too bad, though.
> 2. Sure the effect would be the same, but the code to achieve that effect would be less complicated, and thus easier to maintain.

I promise I will revisit this, but it will definitely take a while. All of these interfaces are internal.

I want to get this in so the December release has the up to date pubsub otel changes.

> 
> > I think it might be a little ugly
> 
> It can be less ugly.
> 
> ```c++
>         auto used_sink = sink;
>         if (!key.empty()) {
>           // Only wrap the sink if there is an ordering key.
>           used_sink = pubsub_internal::SequentialBatchSink::Create(
>               std::move(used_sink));
>         }
>         if (TracingEnabled(opts)) used_sink = MakeTracingBatchSink(std::move(used_sink));
> ```

Duh, that is better. Thanks.